### PR TITLE
Add SFT Recipe

### PR DESCRIPTION
**Pull request by jacklanchantin**
**What does this PR do? Please describe:**
A summary of the change or the issue that is fixed.

- Adds SFT recipe for masked input training.
- Adapts old instruction_finetune recipe for to fit within v0.5 `train` recipe format.

Tested 4 configs using named assets:
```
dataset=alpaca, model=llama_3_2_1b
dataset=alpaca, model=llama_3_2_1b_instruct, chat_mode=true
dataset=gsm8k, model=llama_3_2_1b
dataset=gsm8k, model=llama_3_2_1b_instruct, chat_mode=true
```

Run using:
```
python -m sft /output/dir/ --config-file sft/configs/llama3_2_1b_gsm8k.yaml
````

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [x] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: recipes/lm/sft/default_config.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            compile_options=CompileOptions(fullgraph=True, dynamic=False),
+        )
+    )
+
+    # tokenizer: TokenizerSection = field(
+    #     default_factory=lambda: TokenizerSection(
+    #         config_overrides=LLaMATokenizerConfig(impl="tiktoken", use_eot=True),
+    #         # config_overrides=LLaMATokenizerConfig(impl="hg")
+    #     )
+    # )
```
Comment (jacklanchantin) [inline recipes/lm/sft/default_config.py]: If we specify a local tokenizer path, we have to use impl=tiktoken. but maybe we want to just keep everything as hf?

File: recipes/lm/sft/configs/llama3_2_1b_instruct_alpaca.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+dataset:
+  valid_split: "valid"
+  batch_size: 16
+  chat_mode: true
+  config_overrides:
+    path: "/checkpoint/ram/jacklanchantin/data/alpaca/"
```
Comment (jacklanchantin) [inline recipes/lm/sft/configs/llama3_2_1b_instruct_alpaca.yaml]: need to remove these paths. how should we set default? new dataset card?

File: recipes/lm/sft/dataset.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    proportion to their weights.
+    """
+
+    source_encode_mode: str = "prompt"
+    """The tokenizer mode to encode the source text."""
+
+    target_encode_mode: str = "prompt_response"
+    """The tokenizer mode to encode the target text."""
+
+    chat_mode: bool = False
```
Comment (jacklanchantin) [inline recipes/lm/sft/dataset.py]: maybe we don't need this class anymore and it's all just in [Class DataReadOptions](https://github.com/facebookresearch/fairseq2/pull/1271/files#diff-fc9871ae1e62201b728f966210cf78f11734a2d773bc51133a3bc955e20a6f98R48)?

File: recipes/lm/sft/dataset.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+LM_SFT_DATASET: Final = "lm_sft"
+
+
+@dataclass(kw_only=True)
+class DataReadOptions:
+    batching: Batching = field(default_factory=lambda: StaticBatching(1))
+    """The batching strategy for returned examples."""
+
+    example_shuffle_window: int = 1
```
Comment (uralik) [inline recipes/lm/sft/dataset.py]: lets do true shuffling by default?

File: recipes/lm/sft/dataset.py (hunk 3 above)
Comment (uralik) [reply recipes/lm/sft/dataset.py]: yep I think this can all go there within standalone recipe now

File: recipes/lm/sft/config.py
Comment (uralik) [inline recipes/lm/sft/config.py:1]: shall this be just 'config.py' ?

File: recipes/lm/sft/default_config.py (hunk 1 above)
Comment (uralik) [reply recipes/lm/sft/default_config.py]: yes, default impl=hf makes sense, since this is how we would expect usage with chat mode, + usage with other model like qwen

File: recipes/lm/sft/dataset.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    LengthBatching,
+    StaticBatching,
+    load_files_and_weights,
+)
+
+# TODO: FIX, INFER
+npc = 10
+
+
+LM_SFT_DATASET: Final = "lm_sft"
```
Comment (jacklanchantin) [inline recipes/lm/sft/dataset.py]: change to LM_SFT_PADDED_DATASET "lm_sft_padded"

File: recipes/lm/sft/dataset.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+        for split, (files, weights) in splits.items():
+            if len(files) != len(weights):
+                raise ValueError(
+                    f"The lengths of the file and weight lists of the '{split}' split must match, but they are {len(files)} and {len(weights)} instead."
+                )
+
+        self._splits = splits
+
+    def _read_jsonl(self, path: Path, tokenizer: Tokenizer) -> DataPipelineBuilder:
```
Comment (jacklanchantin) [inline recipes/lm/sft/dataset.py:145]: change to how read_jsonl is done in train recipe

File: recipes/lm/sft/default_config.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    """If True, dataset jsonl must have 'chat' field with openai-like messages List[Dict] entries"""
+
+    min_seq_len: int = 1
+    """The minimum sequence length."""
+
+    max_seq_len: int = 8192
+    """The maximum sequence length."""
+
+    max_num_tokens: int = 8192 * 2
+    """The maximum number of tokens per batch."""
```
Comment (jacklanchantin) [inline recipes/lm/sft/default_config.py]: specify that this is excluding padding tokens

File: recipes/lm/sft/dataset.py (hunk 4 above)
Comment (jacklanchantin) [reply recipes/lm/sft/dataset.py]: changed both of these:

```
example_shuffle_window: int = 0
batch_shuffle_window: int = 0
```

File: recipes/lm/sft/config.py
Comment (jacklanchantin) [reply recipes/lm/sft/config.py:1]: i was basing it off of this https://github.com/facebookresearch/fairseq2/tree/balioglu/wav2vec2_example/recipes/wav2vec2/asr

File: recipes/lm/sft/dataset.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+        # file_world_size = gangs.dp.size
+
+        # if len(files) < file_world_size:
+        #     raise ValueError(
+        #         "The number of dataset files must be greater than or equal to the number of world size."
+        #     )
+
+        # if file_world_size > 1:
+        #     builder.shard(file_rank, file_world_size, allow_uneven=True)
```
Comment (jacklanchantin) [inline recipes/lm/sft/dataset.py]: do we need this? it was used in train recipe, but i don't think we need to enforce data chunks with sft

File: recipes/lm/sft/dataset.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+from .utils import (
+    Batching,
+    DatasetLoadError,
+    LengthBatching,
+    StaticBatching,
+    load_files_and_weights,
+)
+
+# TODO: FIX, INFER
+npc = 10
```
Comment (uralik) [inline recipes/lm/sft/dataset.py]: we have npc in the config below, so this can be removed if we use npc from dataset config

File: recipes/lm/sft/dataset.py (hunk 8 above)
Comment (uralik) [reply recipes/lm/sft/dataset.py]: yes we dont need it, the code
`builder.shard(gangs.dp.rank, gangs.dp.size, allow_uneven=True)` below executes the sharding logic across data workers

File: recipes/lm/sft/config.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    """
+
+    max_num_valid_tokens: int | None = None
+    """The maximum number of tokens per validation batch."""
+
+    example_shuffle_window: int = 10_000
+    """The size of the sliding window for shuffling examples."""
+
+    batch_shuffle_window: int = 1000
+    """The size of the sliding window for shuffling batches."""
```
Comment (uralik) [inline recipes/lm/sft/config.py:71]: we have different defaults here comparing to dataset, lets also set these to 0?

File: recipes/lm/sft/default_config.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    """The encode mode for the target, determines what special tokens to add."""
+
+    chat_mode: bool = False
+    """If True, dataset jsonl must have 'chat' field with openai-like messages List[Dict] entries"""
+
+    min_seq_len: int = 1
+    """The minimum sequence length."""
+
+    max_seq_len: int = 8192
+    """The maximum sequence length."""
```
Comment (uralik) [inline recipes/lm/sft/default_config.py]:
```suggestion
    """The maximum sequence length. NOTE: longer sequences are dropped from the training."""
```

File: recipes/lm/sft/config.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    )
+
+    lr_scheduler: LRSchedulerSection | None = field(
+        default_factory=lambda: LRSchedulerSection(
+            name=COSINE_ANNEALING_LR,
+            config=CosineAnnealingLRConfig(
+                num_warmup_steps=2000, start_lr=1e-30, final_lr_scale=0.01
+            ),
+        ),
+    )
```
Comment (uralik) [inline recipes/lm/sft/config.py:120]: we didnt do such warmups in SFT before right? probably we should disable warmup by default?

File: recipes/lm/sft/config.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    regime: RegimeSection = field(
+        default_factory=lambda: RegimeSection(
+            num_steps=48_000,
+            validate_every_n_steps=4000,
+            checkpoint_every_n_steps=1000,
+            keep_last_n_checkpoints=3,
+            keep_checkpoint_every_n_steps=4000,
+            publish_metrics_every_n_steps=10,
+        )
+    )
```
Comment (uralik) [inline recipes/lm/sft/config.py:133]: lets set args that we typically use? to save HF model, and also this new arg that allows to keep only the last step full train state, but not for older ckpts

File: recipes/lm/sft/default_config.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+@dataclass(kw_only=True)
+class InstructionFinetuneDatasetSection(DatasetSection):
+    # name: str = "foo"  # FIXME: change!
+
+    # family: str = GENERIC_INSTRUCTION_DATASET_FAMILY
+
+    path: str | None = None
+
+    train_split: str = "train"  # FIXME (jacklanchantin)not sure what this should be
```
Comment (uralik) [inline recipes/lm/sft/default_config.py]: can remove comments now and we can set default names here for gsm8k dataset?

File: recipes/lm/sft/default_config.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    ModelSection,
+    OptimizerSection,
+    RegimeSection,
+    TokenizerSection,
+    TorchConfig,
+    TrainerSection,
+)
+from .dataset import LM_SFT_PADDED_DATASET, LMSFTDatasetConfig
+
+GENERIC_INSTRUCTION_DATASET_FAMILY: Final = "generic_instruction"
```
Comment (uralik) [inline recipes/lm/sft/default_config.py]: this is not used right? we have other family name in the dataset.py

Comment (uralik) [reply recipes/lm/sft/default_config.py]: the LM_SFT_PADDED_DATASET

File: recipes/lm/sft/recipe.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+        unit = LMSFTUnit(context.model)
+
+        dataset = context.default_dataset.as_(LMSFTDataset)
+
+        seed = config.common.seed
+        device = get_default_device()
+
+        if device.type == "cuda":
+            torch.cuda.set_device(device)
```
Comment (cbalioglu) [inline recipes/lm/sft/recipe.py]: This is not really needed. fairseq2 already sets the device before calling `create_trainer`. Any particular reason for calling this here?

File: recipes/lm/sft/recipe.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+
+        rng_bag = RngBag.from_device_defaults(CPU, device)
+
+        rng_bag.manual_seed(seed)
```
Comment (cbalioglu) [inline recipes/lm/sft/recipe.py]: This is also not needed. fairseq2 calls `manual_seed(config.common.seed)` before calling `create_trainer`.

File: recipes/lm/sft/recipe.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+    @override
+    def create_trainer(self, context: RecipeContext) -> Trainer:
+        config = context.config.as_(LMSFTConfig)
+
+        unit = LMSFTUnit(context.model)
+
+        dataset = context.default_dataset.as_(LMSFTDataset)
+
+        seed = config.common.seed
+        device = get_default_device()
```
Comment (cbalioglu) [inline recipes/lm/sft/recipe.py]: I think you don't need `device` (see my comments below), but in case needed, a cleaner way is `context.gangs.root.device`.

File: recipes/lm/sft/default_config.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    TokenizerSection,
+    TorchConfig,
+    TrainerSection,
+)
+
+from .dataset import LM_SFT_PADDED_DATASET, LMSFTDatasetConfig
+
+
+@dataclass(kw_only=True)
+class InstructionFinetuneDatasetSection(DatasetSection):
```
Comment (cbalioglu) [inline recipes/lm/sft/default_config.py]: Should we name this `LMSFTDatasetSection`?

**Review by cbalioglu — APPROVED** [review]
Left a few mostly nit comments. Overall looks good to me. Thanks a lot for updating this recipe!

I think one follow-up improvement we should make is revising the dataset implementation to a packed variant. This is on me (or one of the fs2 team members) though.

I am stamping the PR, but would be great if you could check/address the few comments I left.

File: recipes/lm/sft/README.md [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+
+```bash
+# Run with a pre-configured Llama setup
+python -m fairseq2.recipes.lm.sft --config-name recipes/lm/sft/configs/llama3_2_1b_gsm8k.yaml
+
+# Run with a pre-configured Qwen setup
+python -m fairseq2.recipes.lm.sft --config-name recipes/lm/sft/configs/qwen3_4b_gsm8k.yaml
+
+# Run with custom config file")
+python -m fairseq2.recipes.lm.sft --config-file path/to/your/config.yaml
```
Comment (cbalioglu) [inline recipes/lm/sft/README.md]: I think you don't need "fairseq2" here (and in other examples) right? It ideally should work with

```
python -m recipes.lm.sft ...
```

File: src/fairseq2/models/llama/tokenizer.py [hunk 21]
```diff
@@ -191,7 +191,7 @@ def vocab_info(self) -> VocabularyInfo:
 
 @dataclass
 class LLaMATokenizerConfig:
-    impl: Literal["sp", "tiktoken", "hg"] = "sp"
+    impl: Literal["sp", "tiktoken", "hg"] = "hg"
```
Comment (cbalioglu) [inline src/fairseq2/models/llama/tokenizer.py]: Just catching this up. Do you mind keeping this as `sp` so that we don't inadvertently break some dependent libraries? I think as long as we set "hg" as a "model_config" in the base LLaMA-3 asset card, this is not really needed anyways.

File: src/fairseq2/data/__init__.py [hunk 22]
```diff
@@ -0,0 +1,7 @@
+from fairseq2.data.data_pipeline import CollateOptionsOverride as CollateOptionsOverride
+from fairseq2.data.data_pipeline import Collater as Collater
+from fairseq2.data.data_pipeline import DataPipelineBuilder as DataPipelineBuilder
+from fairseq2.data.data_pipeline import SequenceData as SequenceData
+from fairseq2.data.data_pipeline import create_bucket_sizes as create_bucket_sizes
+
+__all__: list[str] = []
```
Comment (cbalioglu) [inline src/fairseq2/data/__init__.py]: Nit: no need for `__all__` here. This can be removed.
